src/realm/object-store/sync/sync_session.hpp

```diff
@@ -216,6 +216,10 @@ class SyncSession : public std::enable_shared_from_this<SyncSession> {
     // longer be open on behalf of it.
     void shutdown_and_wait() REQUIRES(!m_state_mutex, !m_connection_state_mutex);
 
+    // Get notified asynchronously when the synchronization session (sync::Session) is shut down and
+    // the Realm file is no longer open on behalf of it.
+    void shutdown(CloseCallback&&) REQUIRES(!m_state_mutex, !m_connection_state_mutex);
```

Comment:
you could just have this return a `Future<void>` and then we'd get some error handling here for free?

Reply:
I like the idea of using a future, but I guess a `Future<Status>` makes more sense?

Reply:
You don't need `Future<Status>` it already has error handling built in and will return a valid Status object if an error occurs:
```
auto pf = util::make_promise_future<void>();
. . .
pf.future.get(); // Throws an Exception with the Status object on error
// or 
auto result = pf.future.get_no_throw(); // Returns a StatusWith<> object that also contains a Status object
if (!result.is_ok()) {
	printf(result.get_status());
}
// if you want the value, call result.get_value()
```

Reply:
On a second thought, I think I cannot use a future here since it's going to block the caller and that's what I want to avoid.

Reply:
Forgot about `get_async` :) Futures are the future!